### ISSUE TITLE
SimPoints/Sampling methodology?

### ISSUE THREAD

**Issue by austinharris** [darchr/gem5art#106]
I was wondering if there are plans to add a SimPoints or SMARTS style methodology to gem5art? 

In particular (for SimPoints), how would one account for the mismatch in generating SimPoints on a normal machine and then running them in FS mode where the instructions include the kernel and other processes? Is there a way in gem5 to only count instructions for a certain PID, etc?

**Comment by powerjg** [comment]
Thanks for bringing this up. I think this is a very important issue! Though, I don't think my answer will be very satisfying :).

Unfortunately, this isn't on the roadmap right now. However, I (personally) think that sampling methodologies are a very important part of how we're going to continue to improve simulator performance.

I have had success with simple "random" sampling methodologies and gem5. This isn't implemented in gem5art, but i can see how to do this pretty simply. It should be a straightforward extension of the `gem5Run` object which allows for multiple stats dumps and output directories.

There are two main problems with SimPoints (and SMARTS, I think, though I need to study that work again).
1. It will only work for single-threaded applications. When anything except a single-threaded application runs, you won't have determinism because the execution *timing* of one thread can affect the execution *path* of another. You can generalize this with I/O as well.
2. Getting to your question: for FS mode, *general instructions* isn't a good measure of *application* progress. You could have kernel instructions, etc. that are also executed. You need this underlying idea of *application* progress for SimPoints or SMARTS to work.

To address 2), you could probably extend gem5 to count only the instructions for a certain process (e.g., only increment the instruction count when the CR3 is a certain value), but you would have to implement some way to communicate *which* process you care about. This is possible, but not easy. Then, you would still have to contend with things like a different glibc version which could cause divergence between gem5 and hardware.

I *do* have some ideas to address these problems :). However, I haven't had the time/resources to work on it. If this is something that is interesting to you, I'd be happy to chat further!

**Comment by austinharris** [comment]
Thanks for your response Jason!

Right now I am fine with just doing single-threaded applications for the ones that need SimPoints. Basically, I am in agreement with you that FS is much more appropriate for research... but it is still currently still a checkbox for reviewers to have the "standard" methodology for SPEC2017 (specspeed with 1 thread+SimPoints+reference inputs appears to be most popular).

One idea I had for problem 2) is to boot linux with isolcpus=X and then launch spec on X and do the simpoint instruction counting on that core only. Do you think this is reasonable enough?

I am actually using qemu with kvm and valgrind to boot the exact same kernel and disk image when I generate basic block vectors, so I don't think I have a problem there w.r.t glibc, etc.

I would be happy to discuss sometime, but with my current timeline I may have to just do SE mode for spec only. I definitely think this is a vital step in getting the community to embrace full-system simulations though.

**Comment by powerjg** [comment]
As a reviewer, I'm not sure I agree that "specspeed with 1 thread+SimPoints+reference inputs" is what I'm looking for. But then again, I'm probably a unique reviewer :D.

I was not familiar with `isolcpu`... that's a very interesting idea! That *may* work for FS, assuming there's no I/O on that thread. 

Rather than qemu+kvm, why not gem5+kvm! Assuming the BBV is based on virtual address, this will probably work fine. If it's based on dynamic instruction number, this almost certainly won't work. I have no reason to think the execution of an application will be deterministic when it's running on top of an OS.

My advice would be that if you are going to use SimPoints with single-threaded SPEC you should use SE mode. But, it really depends on the question you're asking whether SE vs FS mode is the right approach. For example, if you are proposing a new TLB design, SE mode would *not* be appropriate (it doesn't model TLB misses). However, if you're proposing a new branch predictor which targets user-mode applications, then maybe SE mode would give the same results as FS mode.

Then again, all of this truly is speculation (pun intended). We need a good study on simulator accuracy to show whether or not FS vs SE mode matters or if you can use BBV generated from one mode with another, etc. *Someone* should do this study ;)

**Comment by austinharris** [comment]
You never know when Reviewer #2 wonders why SPEC is not there, or complains about not reference inputs, or (rightfully so) complains about the fast-forward X and simulate Y methodology 😀. My research is in security, so the standard performance benchmarks are more of a check-mark before moving onto more interesting privacy sensitive applications like databases or web browsers that won't run in SE mode. 

I have tried to use the simpoint generation feature in gem5, but I don't believe that the simpoint probe works with the kvmcpu. If you have some insights there I would certainly be willing to look into it some.

**Comment by powerjg** [comment]
Yeah, the simpoint probe isn't going to work out of the box with KVM. I'm not sure what it would take to get it working. I *think* it's doable, though.
